### PR TITLE
Refactor and return 404 from AsmDocs APIs

### LINKED ISSUE #3117: [BUG]: ARM64 instruction documentation fails to display with a cryptic error
### Describe the bug

Attempting to view the documentation for an ARM64 instruction results in an unhelpful error message: ‘There was an error fetching the documentation for this opcode ([object Object]).’

### Steps to reproduce

0. Visit godbolt.org
0. Select an ARM64 compiler
0. Enter some code
0. In the assembly pane, click on a mnemonic and press Ctrl+F8.

### Expected behavior

- Documentation for the assembly instruction is displayed.
- Failing that, an informative error message is displayed.
- Failing that, the option is disabled entirely.

### Reproduction link

https://gcc.godbolt.org/z/MjnzosWdW

### Screenshots

Screenshot of the error

### Operating System

_No response_

### Browser version

Firefox ESR 78

### PR DESCRIPTION
**Pull request by supergrecko**
Fixes #3117 

Refactors the handler for the AsmDocs API to reduce code duplication.

**Breaking changes proposed**

- API will now return 404. This used to return 200 with "Unknown opcode" which is what led to issue #3117. Following cases has been changed:
  1. archiecture was not found
  2. instruction opcode was not found
- API will now return 406 not acceptable when the accepted content type is not text or json
- There is no longer a `result` field on a 200 request, instead the assembly documentation object is returned.
- There is no longer a `found` field on any request, the HTTP codes indicate whether it was found or not (200 means found)

**New features**

- The frontend will only request the opcode for an instruction set at most once, caching every single outgoing request.
- The "There was an error fetching the documentation for this opcode" has been adjusted to contain what went wrong.

Relevant tests have been added and updated.

### CODE REVIEW

File: static/api/api.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+const request = async <R>(uri: string, options?: RequestInit): Promise<TypedResponse<R>> => fetch(
+    `${window.location.origin}${window.httpRoot}api${uri}`,
+    {
+      ...options,
+      headers: {
+          ...options?.headers,
+          'Accept': 'application/json',
+      }
+    },
+);
```
Comment (supergrecko) [inline static/api/api.ts:42]: I think it's a great idea to strongly type all of our API calls (at least the easy ones like format and asm) here with interfaces in /static/api/[component].interfaces.ts

File: lib/handlers/assembly-documentation/router.js [hunk 2]
```diff
@@ -22,37 +22,33 @@
...
+export const setup = (router) => router.get('/asm/:arch/:opcode', (req, res) => {
+    if (MAX_STATIC_AGE > 0) {
+        res.setHeader('Cache-Control', `public, max-age=${MAX_STATIC_AGE}`);
+    }
+    const handler = ASSEMBLY_DOCUMENTATION_HANDLERS[req.params.arch];
+    if (handler !== undefined) {
+        return handler.handle(req, res);
     }
-}
+    res.status(404).send({ error: 'Unknown architecture '});
```
Comment (mattgodbolt) [inline lib/handlers/assembly-documentation/router.js:53]: 👍🏻

File: lib/handlers/base-assembly-documentation-handler.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ * @typedef {object} AssemblyInstructionInfo
```
Comment (mattgodbolt) [inline lib/handlers/base-assembly-documentation-handler.js:26]: Thanks - learning so much from reading this PR.

File: lib/handlers/base-assembly-documentation-handler.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * Implementors should not have to override this.
+     *
+     * @param {e.Request} request
+     * @param {e.Response} response
+     */
+    handle(request, response) {
+        // If the request had no opcode parameter, we should fail. This assumes
+        // no assembly language has a __unknown_opcode instruction.
+        const instruction = request.params.opcode.toUpperCase() || '__UNKNOWN_OPCODE';
```
Comment (mattgodbolt) [inline lib/handlers/base-assembly-documentation-handler.js:58]: If they had no opcode, wouldn't the `toUpperCase()` fail ? Does this need to be
```suggestion
        const instruction = (request.params.opcode || '__UNKNOWN_OPCODE').toUpperCase();
```

File: lib/handlers/base-assembly-documentation-handler.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        const information = this.getInstructionInformation(instruction);
+        if (information === null) {
+            return response.status(404).send({ error: 'Unknown opcode' });
+        }
+        // Accept either JSON or Plaintext Content-Type
+        const requestedContentType = request.accepts(['text', 'json']);
+        switch (requestedContentType) {
+            case 'text': response.send(information.html); break;
+            case 'json': response.send(information); break;
+            default: response.status(406).send({ error: 'Not Acceptable' }); break;
```
Comment (mattgodbolt) [inline lib/handlers/base-assembly-documentation-handler.js:68]: TIL 406 :)

File: static/api/api.ts (hunk 1 above)
Comment (mattgodbolt) [reply static/api/api.ts:42]: Sounds great to me!

File: static/api/assemblydocumentation.interfaces.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+export interface AssemblyDocumentationRequest {
+    /** Specifies which instruction set to look for */
+    instructionSet: 'amd64' | 'arm32' | 'java';
```
Comment (mattgodbolt) [inline static/api/assemblydocumentation.interfaces.ts:27]: Neat. Another place to remember to update though.

File: static/panes/compiler.js [hunk 7]
```diff
@@ -2228,24 +2229,26 @@ function getAsmInfo(opcode, instructionSet) {
...
-            success: function (result) {
-                OpcodeCache.set(cacheName, result);
-                resolve(result.found ? result.result : null);
-            },
-            error: function (result) {
-                reject(result);
-            },
-            cache: true,
-        });
+        getAssemblyDocumentation({ opcode: opcode, instructionSet: instructionSet })
```
Comment (mattgodbolt) [inline static/panes/compiler.js:2238]: So much nicer!

File: static/api/api.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    json(): Promise<R>
+}
+
+/** Lightweight fetch() wrapper for CE API urls */
+const request = async <R>(uri: string, options?: RequestInit): Promise<TypedResponse<R>> => fetch(
+    `${window.location.origin}${window.httpRoot}api${uri}`,
+    {
+      ...options,
+      headers: {
+          ...options?.headers,
```
Comment (mattgodbolt) [inline static/api/api.ts:38]: nice :)

File: test/handlers/api-tests.js [hunk 9]
```diff
@@ -173,7 +173,6 @@ describe('API handling', () => {
             .then(res => {
                 res.should.have.status(200);
                 res.should.be.json;
-                res.body.found.should.be.true;
```
Comment (mattgodbolt) [inline test/handlers/api-tests.js:176]: :+1:

**Review by mattgodbolt — APPROVED** [review]
Looks great! Thanks so much!

File: lib/handlers/base-assembly-documentation-handler.js (hunk 4 above)
Comment (supergrecko) [reply lib/handlers/base-assembly-documentation-handler.js:58]: You're absolutely correct! Thanks!

**Pull request merged**